packages/x-date-pickers/src/internals/hooks/useField/buildSectionsFromFormat.ts

```diff
@@ -202,56 +202,48 @@ const buildSections = <TDate extends PickerValidDate>(
   const sections: FieldSection[] = [];
   let startSeparator: string = '';
 
-  // This RegExp test if the beginning of a string corresponds to a supported token
-  const isTokenStartRegExp = new RegExp(
-    `^(${Object.keys(utils.formatTokenMap)
-      .sort((a, b) => b.length - a.length) // Sort to put longest word first
-      .join('|')})`,
-    'g', // used to get access to lastIndex state
-  );
+  // This RegExp tests if the beginning of a string corresponds to a supported token
```

Comment:
I reworked this method because it was duplicating the 1st character of the 2nd token (once inside the token and once as a separator).

Since we catch the whole token in one passage in the loop, we can replace the `for` with a `while` and stop populating `currentTokenValue` incrementaly. This makes the whole behavior simple IMHO.